### PR TITLE
SKU SDK should be able to handle more than one environment

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/20984

Removed Skus enviroment flag, production env is used by default in official builds and development in others. Latest successfully authorized env becomes active. Added some tests.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Test jumping between envs and enable/disable vpn

### CODE REVIEW

File: browser/brave_vpn/brave_vpn_service_factory.cc [hunk 1]
```diff
@@ -41,6 +41,19 @@ BraveVpnService* BraveVpnServiceFactory::GetForProfile(Profile* profile) {
       GetInstance()->GetServiceForBrowserContext(profile, true));
 }
 
+// static
+#if !BUILDFLAG(IS_ANDROID)
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply browser/brave_vpn/brave_vpn_service_factory.cc:45]: There is no BindInterface for VPN service

File: components/skus/renderer/skus_js_handler.cc [hunk 2]
```diff
@@ -304,6 +313,9 @@ void SkusJSHandler::OnCredentialSummary(
     return;
   }
 
+  auto host = render_frame_->GetWebFrame()->GetSecurityOrigin().Host().Utf8();
+  vpn_service_->LoadPurchasedState(domain);
```
[the remark this answers is not in the document]
Comment (spylogsster) [reply components/skus/renderer/skus_js_handler.cc:315]: discussed in slack

File: components/brave_vpn/brave_vpn_service.cc [hunk 3]
```diff
@@ -1020,18 +1033,24 @@ void BraveVpnService::OnPrepareCredentialsPresentation(
...
   purchased_state_ = state;
-  VLOG(1) << "SetPurchasedState: " << purchased_state_.value();
 
   for (const auto& obs : observers_)
     obs->OnPurchasedStateChanged(purchased_state_.value());
 }
 
+void BraveVpnService::SetCurrentEnvironment(const std::string& env) {
+  current_env_ = env;
+  purchased_state_.reset();
```
Comment (simonhong) [inline components/brave_vpn/brave_vpn_service.cc:1051]: Don't need to notify this purchased state change?

File: components/brave_vpn/BUILD.gn [hunk 4]
```diff
@@ -109,7 +109,10 @@ source_set("unit_tests") {
 
     # Currently, tests are for desktop only (fails on Android).
     if (is_win || is_mac) {
-      sources += [ "brave_vpn_unittest.cc" ]
+      sources += [
+        "brave_vpn_unittest.cc",
+        "skus_utils_unittest.cc",
```
Comment (simonhong) [inline components/brave_vpn/BUILD.gn:114]: I think `skus_utils_unittest.cc` should be included in `//brave/components/skus/browser`?

**Review by simonhong — DISMISSED** [review]
++ for `//brave/browser`, `//brave/components/brave_vpn`

File: components/brave_vpn/BUILD.gn (hunk 4 above)
Comment (spylogsster) [reply components/brave_vpn/BUILD.gn]: done

File: browser/brave_vpn/brave_vpn_service_factory.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/brave_vpn/brave_vpn_service_factory.cc:45]: Ok, this should be handled when we want to use credential & purchased state on Android.

**Comment by kylehickinson** [comment]
@spylogsster So just to clarify, we will no longer need to set an environment to handle development/staging SKUs? Or is there some other integration point? I ask because iOS essentially only ever uses Release builds for a vast majority of our builds so we'll still need some way of switching to dev/staging without building a local copy of core

**Comment by spylogsster** [comment]
> @spylogsster So just to clarify, we will no longer need to set an environment to handle development/staging SKUs? Or is there some other integration point? I ask because iOS essentially only ever uses Release builds for a vast majority of our builds so we'll still need some way of switching to dev/staging without building a local copy of core

yes, in official builds production env os default but you can visit deveopment domains and re-authorize with development credentials

**Review by kylehickinson — APPROVED** [review]
iOS++

File: components/skus/browser/skus_utils.cc [hunk 5]
```diff
@@ -54,6 +46,19 @@ std::string GetDomain(std::string prefix) {
   return "";
 }
 
+std::string GetEnvironmentForDomain(const std::string& domain) {
+  auto base_domain = GetDomainAndRegistry(
```
Comment (bsclifton) [inline components/skus/browser/skus_utils.cc:50]: Would it make sense to fully qualify this? ex: `net::registry_controlled_domains::GetDomainAndRegistry`

**Review by bsclifton — APPROVED** [review]
Dug deep in here; this is a great solution w/ the unordered map and `GetOrCreateSDK`. I rebased as the old HEAD was having a crash when logging in; did sync/build and it works perfectly

I did testing using dev and staging. You can switch between the windows and then (when wanting to switch environments) click `Refresh Brave VPN`. Following the manage URL is how I verified the correct environment specific URL was being created

Nice work on this! Plus, the mojom will help w/ https://github.com/brave/brave-core/pull/13387

File: components/skus/browser/skus_utils.cc (hunk 5 above)
Comment (spylogsster) [reply components/skus/browser/skus_utils.cc]: done

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**